zilencer/lib/remote_counts.py

```diff
@@ -0,0 +1,85 @@
+import datetime
+
+from django.db import connection
+from django.utils.timezone import now as timezone_now
+from psycopg2.sql import SQL, Literal
+
+from zilencer.models import RemoteInstallationCount, RemoteZulipServer
+
+
+class MissingDataError(Exception):
+    pass
+
+
+def compute_max_monthly_messages(remote_server: RemoteZulipServer) -> int:
+    # Calculate the maximum amount of messages that the server had within a month.
+    # out of the last 3 months.
+
+    # Filtering on property is a placeholder here; it
+    # currently doesn't change the result. Note that we can't
+    # filter on the actual property we care about
+    # ('messages_sent:message_type:day'), because zeros are not
+    # recorded.
+    # For non-legacy servers, we probably want to distinguish not uploading
+    # data vs. uploading minimal billing data (and thus not qualified for the
+    # Starter plan) vs. uploading all the data.
+    if not RemoteInstallationCount.objects.filter(
+        server=remote_server,
+        property="active_users_audit:is_bot:day",
+        end_time__lte=timezone_now() - datetime.timedelta(days=3),
+    ).exists():
+        raise MissingDataError
```

Comment:
In the bulk query this check for the last 3 days of data was built into the single SQL query, but if we're querying about a single server, then it shouldn't be an issue to do a separate first query just for it and it significantly improves readability?

Reply:
Yeah, agreed -- we want to bubble this up as your exception logic does it. The bulk query was just working around the limitations of doing everything in SQL.